### PR TITLE
Ka/resource loader

### PR DESCRIPTION
**Pull request by kalagarraj**
raml-toolkit now loads commerce standard resource files using a custom resource loader

### CODE REVIEW

File: src/exchange-connector/resourceLoader/commerceFileSystemAdapter.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { readFileSync } from "fs-extra";
+
+export class CommerceFileSystemAdapter {
```
[the remark this answers is not in the document]
Comment (kalagarraj) [reply src/exchange-connector/resourceLoader/commerceFileSystemAdapter.ts]: Changed to FileSystemAdapter

File: src/exchange-connector/resourceLoader/commerceStandardsLoader.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import { CommerceFileSystemAdapter } from "./commerceFileSystemAdapter";
+
+const EXCHANGE_MODULES = "exchange_modules/";
+
+export class CommerceStandardsLoader implements amf.resource.ResourceLoader {
+  public workingDir: string;
+  public fsAdapter: CommerceFileSystemAdapter;
+
+  constructor(rootFolder: string) {
+    this.workingDir = rootFolder.split("file://")[1];
```
[the remark this answers is not in the document]
Comment (kalagarraj) [reply src/exchange-connector/resourceLoader/commerceStandardsLoader.ts:19]: Does it work? I am getting typescript error for invalid ternary operator

File: src/exchange-connector/resourceLoader/commerceStandardsLoader.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    if (!this.workingDir) {
+      throw new Error("rootFolder does not contain 'file://' prefix");
+    }
+    this.fsAdapter = new CommerceFileSystemAdapter();
+  }
+
+  accepts(resource: string): boolean {
```
[the remark this answers is not in the document]
Comment (kalagarraj) [reply src/exchange-connector/resourceLoader/commerceStandardsLoader.ts:26]: Done

File: src/exchange-connector/resourceLoader/commerceStandardsLoader.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bredmond-sf) [reply src/exchange-connector/resourceLoader/commerceStandardsLoader.ts:19]: nevermind, I was thinking it would throw for out of bounds, but of javascript doesn't do that!

File: src/exchange-connector/resourceLoader/index.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+"use strict";
+
+export { CommerceStandardsLoader } from "./commerceStandardsLoader";
```
Comment (monkey-jeff) [inline src/exchange-connector/resourceLoader/index.ts:9]: I don't like this name.  It is too specific to something that should be broader.  Can we call it `FatRamlResourceLoader` instead pleasE?

Comment (monkey-jeff) [reply src/exchange-connector/resourceLoader/index.ts:9]: What do you mean subsequent PR?  Why split it?

This change changes what the interface is and breaking changes like that I would prefer to keep to a single PR

Comment (kalagarraj) [reply src/exchange-connector/resourceLoader/index.ts]: just want to have the name change review easier... Anyways changed it in this PR itself...

File: src/exchange-connector/resourceLoader/fatRamlResourceLoader.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import { FileSystemAdapter } from "./fileSystemAdapter";
+
+const EXCHANGE_MODULES = "exchange_modules/";
+
+export class FatRamlResourceLoader implements amf.resource.ResourceLoader {
+  public workingDir: string;
+  public fsAdapter: FileSystemAdapter;
+
+  constructor(rootFolder: string) {
+    this.workingDir = rootFolder.split("file://")[1];
```
Comment (monkey-jeff) [inline src/exchange-connector/resourceLoader/fatRamlResourceLoader.ts:19]: Can we assume that it will always be lowercase here? Or should we toLower it for the split?

File: src/exchange-connector/resourceLoader/fatRamlResourceLoader.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    if (!this.workingDir) {
```
Comment (monkey-jeff) [inline src/exchange-connector/resourceLoader/fatRamlResourceLoader.ts:20]: Is it important that we have the file prefix here?

File: src/exchange-connector/resourceLoader/fatRamlResourceLoader.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      throw new Error("rootFolder does not contain 'file://' prefix");
+    }
+    this.fsAdapter = new FileSystemAdapter();
+  }
+
+  accepts(resource: string): boolean {
+    return resource?.indexOf(EXCHANGE_MODULES) >= 0;
+  }
+
+  fetch(resource: string): Promise<amf.client.remote.Content> {
```
Comment (monkey-jeff) [inline src/exchange-connector/resourceLoader/fatRamlResourceLoader.ts:30]: [NIT] Could simplify by making it async

File: src/exchange-connector/resourceLoader/fatRamlResourceLoader.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    return new Promise((resolve, reject) => {
+      if (resource.indexOf(EXCHANGE_MODULES) >= 0) {
```
Comment (monkey-jeff) [inline src/exchange-connector/resourceLoader/fatRamlResourceLoader.ts:32]: Not really resilient as this would find exchange_modules anywhere in the. path and move it to the top level.  Not sure if that is what you want or if you want just if there is exchange_modeles at the beginning of the resource path instead of anywhere.

File: src/exchange-connector/resourceLoader/fileSystemAdapter.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { readFileSync } from "fs-extra";
+
+export class FileSystemAdapter {
```
Comment (monkey-jeff) [inline src/exchange-connector/resourceLoader/fileSystemAdapter.ts:10]: This seems like an awful lot just to avoid writing
```
readFileSync(path, "utf8");
```

As you are taking one method that is used statically and wrapping it in a class you need to instantiate to use? I would advise against this.

File: src/validator.ts [hunk 10]
```diff
@@ -38,7 +39,13 @@ export async function validateModel(
 export async function parseRaml(
   filename: string
 ): Promise<amf.model.document.BaseUnit> {
-  const parser = new Raml10Parser();
+  const ccStandardResourceLoader = new FatRamlResourceLoader(
```
Comment (monkey-jeff) [inline src/validator.ts:42]: Refactors the class name but then this instantiates it as ccStandardResourceLoader.  I would prefer this to also be called something like fatRamlResourceLoader instead.

File: test/exchange-connector/resourceLoader/fatRamlResourceLoader.test.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+import amf from "amf-client-js";
+
+before(() => {
+  chai.should();
+  chai.use(chaiAsPromised);
+});
+
+describe("Commerce standards resource loader fsAdapter tests", () => {
+  it("Throws error on undefined resource absolute path", () => {
+    const ccStandardsLoader = new FatRamlResourceLoader("file:///rootFolder");
```
Comment (monkey-jeff) [inline test/exchange-connector/resourceLoader/fatRamlResourceLoader.test.ts:23]: Same comment as above

File: test/exchange-connector/resourceLoader/fatRamlResourceLoader.test.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+    return expect(() =>
+      ccStandardsLoader.fsAdapter.readFileSync(undefined)
+    ).to.throw();
+  });
+});
+
+describe("Commerce standards resource loader create tests", () => {
```
Comment (monkey-jeff) [inline test/exchange-connector/resourceLoader/fatRamlResourceLoader.test.ts:31]: Should be renamed to reflect the other change

File: test/exchange-connector/resourceLoader/fatRamlResourceLoader.test.ts [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+    );
+  });
+
+  it("Throws error on null root path", () => {
+    return expect(() => new FatRamlResourceLoader(null)).to.throw(
+      "Cannot read property 'split' of null"
+    );
+  });
+
+  it("Throws error on root path without 'file://' prefix", () => {
```
Comment (monkey-jeff) [inline test/exchange-connector/resourceLoader/fatRamlResourceLoader.test.ts:44]: could use some more edge tests.

Like

```
file://///something
```

```
FILE://
```
etc

Write tests with the intention of breaking your stuff.

File: test/exchange-connector/resourceLoader/fatRamlResourceLoader.test.ts [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+describe("Commerce standards resource loader accepts tests", () => {
+  let standardsLoader;
+
+  before(() => {
+    standardsLoader = new FatRamlResourceLoader("file:///rootFolder");
+  });
+
+  it("Returns true (accepted) for resource path containing 'exchange_modules/'", () => {
+    const accepted = standardsLoader.accepts("exchange_modules/");
```
Comment (monkey-jeff) [inline test/exchange-connector/resourceLoader/fatRamlResourceLoader.test.ts:59]: Should have some more happy path tests here.  Liking having this with other content

File: tsconfig.json [hunk 15]
```diff
@@ -18,8 +18,7 @@
         "noImplicitReturns": true,
         "noImplicitThis": true,
         "preserveConstEnums": true,
-        "outDir": "./lib",
-        "rootDir": "./src"
```
Comment (monkey-jeff) [inline tsconfig.json:22]: Why this change? I played with. this and removing this broke the ability to use the generated package where the tests still worked.  rootDir was here to prevent that from happening.

**Comment by monkey-jeff** [comment]
Build has failed due to coverage levels?

**Comment by kalagarraj** [comment]
> Build has failed due to coverage levels?

All set now. mercury.raml was moved incorrectly in one of the commits
